app/src/main/kotlin/com/hippo/ehviewer/download/DownloadManager.kt

```diff
@@ -530,26 +537,47 @@ object DownloadManager : OnSpiderListener, CoroutineScope {
     }
 
     suspend fun readMetadataFromLocal() {
-        val list = allInfoList.filter { it.pages == 0 || it.simpleTags == null }.parMapNotNull(concurrency = 5) { info ->
+        val galleryMutex = Mutex()
+        val galleryInfoList = mutableListOf<BaseGalleryInfo>()
+        val artistMutex = Mutex()
+        val artistInfoList = mutableListOf<Pair<Long, List<DownloadArtist>>>()
+
+        allInfoList.mapNotNull {
+            val updateGallery = it.pages == 0 || it.simpleTags == null
+            val updateArtist = it.artistInfoList.isEmpty()
+            if (updateGallery || updateArtist) {
+                Triple(it, updateGallery, updateArtist)
+            } else {
+                null
+            }
+        }.parMap(concurrency = 5) { (info, updateGallery, updateArtist) ->
```

Comment:
Consider let `parMap` lambda return `Pair<BaseGalleryInfo, List<DownloadArtist>>` and avoid `Mutex`